Add .navbar-nav-scroll for vertical scrolling of navbar content

mdo opened the pull request:
Fixes #23374.

TODO:

- [x] Fix example

Preview: <https://deploy-preview-32037--twbs-bootstrap.netlify.app/docs/5.0/components/navbar/#scrolling>

mdo committed:
Test out a .navbar-nav-scroll class

mdo committed:
Document it better, remove commented out code

XhmikosR commented on site/content/docs/5.0/components/navbar.md:
```diff
@@ -479,6 +480,59 @@ Also note that **`.sticky-top` uses `position: sticky`, which [isn't fully suppo
 
+## Scrolling
+
+Add `.navbar-nav-scroll` to a `.navbar-nav` to enable vertical scrolling when expanded navbar content gets too tall. By default, scrolling kicks in at `75vh` (or 75% of the viewport height), but you may override that with a local CSS custom property. For example, add `style="--height: 65vh;"` to the `.navbar-nav`.
```
`70%`?

mdo replied:
Are you suggesting we change something to `70%`?

XhmikosR replied:
Isn't the default 70%?

ffoodd commented on site/content/docs/5.0/components/navbar.md:
```diff
@@ -43,7 +44,7 @@ Here's an example of all the sub-components included in a responsive light-theme
       <span class="navbar-toggler-icon"></span>
     </button>
-    <div class="collapse navbar-collapse" id="navbarSupportedContent">
+    <div class="collapse navbar-collapse navbar-scroll" id="navbarSupportedContent">
```
Shouldn't it be `navbar-nav-scroll`?

patrickhlauke commented:
nitpick ... not just vertical scrolling, but horizontal scrolling as well, no?

ffoodd commented:
Just a suggestion here: couldn't we use the [scrolling shadow trick by Lea Verou](https://lea.verou.me/2012/04/background-attachment-local/) here?

Would help to get a visual clue of overflown things.

And maybe some `scroll-snap-*` love too? Don't have a concrete idea but there's probably something to do.

patrickhlauke commented:
> Just a suggestion here: couldn't we use the [scrolling shadow trick by Lea Verou](https://lea.verou.me/2012/04/background-attachment-local/) here?
> 
> Would help to get a visual clue of overflown things.
> 
> And maybe some `scroll-snap-*` love too? Don't have a concrete idea but there's probably something to do.

i'd suggest a new ticket for this. and agree, would be good to provide a visual hint that there's more that can be reached with scrolling (for platforms where the scrollbar is not shown, like mobile). maybe not a shadow, but perhaps a subtle arrow or similar if that's still possible to do with pure CSS

mdo committed:
Rename local variable, add more to docs, fix value inconsistency, move class from first demo, link to the demo from up top

mdo commented:
Rebased and updated this PR to prevent the horizontal scrolling issue on expanded navbars. Also documented the potential issue with using `overflow-y` and clipping through `overflow-x` being automatically set by browsers. I think this is good to go and could be nice to include in both v5 and v4.6.

Can y'all take another look?

mdo committed:
Fix dupe ID

XhmikosR commented:
@mdo it seems there are still overflow in certain display widths

![image](https://user-images.githubusercontent.com/349621/104116407-ef132400-5320-11eb-88de-7d020a2e1dea.png)

But it seems to affect the other snippets too. That being said, I'm having trouble visualizing the scroll part here; I don't see the ability to horizontally scroll at all.

patrickhlauke committed:
Merge branch 'main' into navbar-scroll

ffoodd commented:
Like @XhmikosR said, even duplicating some nav items it still overflows horizontally. Am I missing something?

mdo commented:
> That being said, I'm having trouble visualizing the scroll part here; I don't see the ability to horizontally scroll at all.

This is for vertical scroll, not horizontal. On the expanded navbars (that is to say, the horizontal ones) the intent is that you don't see anything different than other navbars. Resetting to overflow visible might not be the right move though (maybe unset or initial?), but I was aiming for something that can be backported to v4.

XhmikosR commented:
TBH I'm not sure about the feature itself, but if you all think it's something useful, that's fine.

BTW the scrollbar seems to overlap here with the button (FF on Windows x64)

<img src="https://user-images.githubusercontent.com/349621/104429873-c09a8080-558e-11eb-8830-95fd142b085d.png" alt="image" width="400">

mdo committed:
Clean up language, add adjust demo, add update margin spacers

mdo commented:
Scrollbar only overlapped the button focus, but I've tweaked the demo to prevent that. Also updated the docs verbiage a bit, so should be good to go now.

XhmikosR committed:
Merge branch 'main' into navbar-scroll

baszero commented:
Is this available in Bootstrap 4.5 ?
